Fix Indexer function failing to communicate with GKE/Elastic

# Task

Indexer always fails with:

```
org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938) at java.base/java.lang.Thread.run(Thread.java:829) Caused by: java.lang.RuntimeException: Error authenticating against GKE at com.moneycol.indexer.infra.connectivity.gke.GkeClient.getServiceDetails(GkeClient.java:56) at com.moneycol.indexer.infra.connectivity.gke.InGkeElasticSearchDiscoveryClient.obtainEndpoint(InGkeElasticSearchDiscoveryClient.java:24) at com.moneycol.indexer.indexing.index.ElasticClientFactory.elasticSearchClient(ElasticClientFactory.java:21) at com.moneycol.indexer.indexing.index.$ElasticClientFactory$ElasticSearchClient0$Definition.build(Unknown Source) at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2336) ... 58 more Caused by: java.lang.RuntimeException: Cannot find any internal address in cluster at com.moneycol.indexer.infra.connectivity.gke.GkeClient.getNodeIpByType(GkeClient.java:87) at com.moneycol.indexer.infra.connectivity.gke.GkeClient.getServiceDetails(GkeClient.java:48) ...
```

Probably the timing of indexer finishing: 

```
2022-08-14 12:12:59.526 BST
```

Does not match when the GKE cluster is up

# Acceptance criteria

- [x] Indexer is able to communicate with GKE and get the data into Elasticsearch

# Definition of Done
1. - [x] Current tests pass
2. - [x] New tests added
3. - [x] Pull Request and self review
4. - [x] Clean build

--- a/data-collector/crawler/infra/terraform/data-collector/backend.tf
+++ b/data-collector/crawler/infra/terraform/data-collector/backend.tf
@@ -1,6 +1,6 @@
 terraform {
   backend "gcs" {
-    bucket  = "moneycol-terraform"
-    prefix  = "dev/data-collector"
+    bucket = "moneycol-terraform"
+    prefix = "dev/data-collector"
   }
 }

--- a/data-collector/crawler/infra/terraform/data-collector/main.tf
+++ b/data-collector/crawler/infra/terraform/data-collector/main.tf
@@ -1,8 +1,8 @@
 resource "google_service_account" "data_collector_main_service_account" {
-    account_id   = "data-collector"
-    description  = "Data Collector (aka Crawler)"
-    disabled     = false
-    display_name = "data-collector"
+  account_id   = "data-collector"
+  description  = "Data Collector (aka Crawler)"
+  disabled     = false
+  display_name = "data-collector"
 }
 
 resource "google_project_iam_custom_role" "data_collector_role" {
@@ -10,13 +10,13 @@ resource "google_project_iam_custom_role" "data_collector_role" {
   title       = "Data Collector"
   description = "Data Collector"
   permissions = [
-      "pubsub.topics.publish",
-      "storage.buckets.create",
-      "storage.objects.create",
-      "storage.objects.delete",
-      "storage.objects.get",
-      "storage.objects.list"
-    ]
+    "pubsub.topics.publish",
+    "storage.buckets.create",
+    "storage.objects.create",
+    "storage.objects.delete",
+    "storage.objects.get",
+    "storage.objects.list"
+  ]
 }
 
 resource "google_project_iam_binding" "data_collector_sa_project_iam_binding" {
@@ -32,10 +32,10 @@ resource "google_service_account_key" "data_collector_sa_key" {
 }
 
 resource "google_service_account" "gke_resize_main_service_account" {
-    account_id   = "gke-resize"
-    description  = "GKE resize"
-    disabled     = false
-    display_name = "gke-resize"
+  account_id   = "gke-resize"
+  description  = "GKE resize"
+  disabled     = false
+  display_name = "gke-resize"
 }
 
 resource "google_project_iam_custom_role" "gke_resize_role" {
@@ -50,7 +50,7 @@ resource "google_project_iam_custom_role" "gke_resize_role" {
     "container.pods.list",
     "container.services.get",
     "cloudfunctions.functions.invoke"
-    ]
+  ]
 }
 
 resource "google_project_iam_binding" "gke_resize_sa_project_iam_binding" {
@@ -75,18 +75,31 @@ resource "google_cloud_scheduler_job" "start_crawler_job" {
   http_target {
     http_method = "POST"
     uri         = var.crawler_resize_uri
-    body = base64encode(var.crawler_resize_payload)
-    headers = { 
-        "Content-Type" = "application/json"
-        "User-Agent" = "Google-Cloud-Scheduler"
+    body        = base64encode(var.crawler_resize_payload)
+    headers = {
+      "Content-Type" = "application/json"
+      "User-Agent"   = "Google-Cloud-Scheduler"
     }
 
     oidc_token {
       service_account_email = google_service_account.gke_resize_main_service_account.email
     }
   }
+
+  retry_config {
+    max_backoff_duration = "3600s"
+    max_doublings        = 5
+    max_retry_duration   = "0s"
+    min_backoff_duration = "5s"
+    retry_count          = 0
+  }
 }
 
+# Add GKE shrink / expand here
+
+
+# for this to work it's needed to auth against GKE
+# gcloud container clusters get-credentials cluster-dev2 --zone europe-west1-b --project moneycol
 resource "kubernetes_secret" "data_collector_key_secret" {
   metadata {
     name = "data-collector-key"
@@ -105,9 +118,9 @@ resource "kubernetes_secret" "data_collector_key_secret" {
 # the function resizes the GKE indexer node pool back to 0
 
 locals {
-    index_resizer_function_name = var.index_resizer_function_name
-    index_resizer_function_code_path = var.index_resizer_function_code_path
-    function_trigger_topic_name = var.crawling_done_pubsub_topic
+  index_resizer_function_name      = var.index_resizer_function_name
+  index_resizer_function_code_path = var.index_resizer_function_code_path
+  function_trigger_topic_name      = var.crawling_done_pubsub_topic
 }
 
 data "google_pubsub_topic" "function_topic" {
@@ -120,9 +133,9 @@ resource "google_storage_bucket" "functions_bucket" {
 }
 
 data "archive_file" "function_archive" {
-  type         = "zip"
-  source_dir   = local.index_resizer_function_code_path
-  output_path  = "${local.index_resizer_function_name}.zip"
+  type        = "zip"
+  source_dir  = local.index_resizer_function_code_path
+  output_path = "${local.index_resizer_function_name}.zip"
 }
 
 resource "google_storage_bucket_object" "archive" {
@@ -132,21 +145,21 @@ resource "google_storage_bucket_object" "archive" {
 }
 
 resource "google_cloudfunctions_function" "resize_function" {
-  name        = "${local.index_resizer_function_name}"
-  description = "Resize indexer node pool to 0"
-  runtime     = "nodejs14"
+  name                  = local.index_resizer_function_name
+  description           = "Resize indexer node pool to 0"
+  runtime               = "nodejs14"
   available_memory_mb   = 128
   source_archive_bucket = google_storage_bucket.functions_bucket.name
   source_archive_object = google_storage_bucket_object.archive.name
   timeout               = 60
-  entry_point = "resizeCluster"
+  entry_point           = "resizeCluster"
 
-  event_trigger  {
+  event_trigger {
     event_type = "providers/cloud.pubsub/eventTypes/topic.publish"
-    resource   = "${data.google_pubsub_topic.function_topic.name}"
+    resource   = data.google_pubsub_topic.function_topic.name
   }
 
   lifecycle {
-      create_before_destroy = true
+    create_before_destroy = true
   }
 }

--- a/data-collector/crawler/infra/terraform/data-collector/pubsub.tf
+++ b/data-collector/crawler/infra/terraform/data-collector/pubsub.tf
@@ -1,0 +1,21 @@
+resource "google_pubsub_topic" "indexer_sink_pubsub_topic" {
+  name = var.indexer_sink_pubsub_topic
+
+  labels = {
+    service = "data-collector"
+  }
+
+  message_retention_duration = "86600s"
+}
+
+resource "google_pubsub_subscription" "indexer_sink_pubsub_subscription" {
+  name = var.indexer_sink_pubsub_topic
+  topic = google_pubsub_topic.indexer_sink_pubsub_topic.name
+  ack_deadline_seconds = 600
+
+  labels = {
+    service = "data-collector"
+  }
+
+  message_retention_duration = "86600s"
+}

--- a/data-collector/crawler/infra/terraform/data-collector/pubsub.tf
+++ b/data-collector/crawler/infra/terraform/data-collector/pubsub.tf
@@ -9,8 +9,8 @@ resource "google_pubsub_topic" "indexer_sink_pubsub_topic" {
 }
 
 resource "google_pubsub_subscription" "indexer_sink_pubsub_subscription" {
-  name = var.indexer_sink_pubsub_topic
-  topic = google_pubsub_topic.indexer_sink_pubsub_topic.name
+  name                 = var.indexer_sink_pubsub_topic
+  topic                = google_pubsub_topic.indexer_sink_pubsub_topic.name
   ack_deadline_seconds = 600
 
   labels = {

--- a/data-collector/crawler/infra/terraform/data-collector/variables.tf
+++ b/data-collector/crawler/infra/terraform/data-collector/variables.tf
@@ -30,11 +30,17 @@ variable crawler_resize_payload {
 }
 
 variable crawler_resize_job_schedule {
-    default = "0 1 * * 0"
+    # Indexing only happens on 1st of each month for now
+    default = "0 1 1 * *"
     description = "The cron schedule to resize the indexing node pool to 1"
 }
 
 variable crawling_done_pubsub_topic {
     default = "dev.crawler.events"
     description = "The resizer function (back to 0, crawling done) is triggered when message is published to this topic"
+}
+
+variable indexer_sink_pubsub_topic {
+    default = "dev.moneycol.indexer.sink"
+    description = "The Pub/Sub topic/subscription where data ready to index is pushed by batcher/worker functions"
 }

--- a/data-collector/crawler/infra/terraform/data-collector/variables.tf
+++ b/data-collector/crawler/infra/terraform/data-collector/variables.tf
@@ -1,46 +1,46 @@
-variable gcp_project {
-    default = "moneycol"
-    description = "GCP project"
+variable "gcp_project" {
+  default     = "moneycol"
+  description = "GCP project"
 }
 
-variable region {
-    default = "europe-west1"
-    description = "Region of the Moneycol GCP project "
+variable "region" {
+  default     = "europe-west1"
+  description = "Region of the Moneycol GCP project "
 }
 
-variable index_resizer_function_name {
-    default = "resize-down-indexer"
-    description = "The Cloud Function that resizes the node pool in GKE performing crawling (indexing-pool)"
+variable "index_resizer_function_name" {
+  default     = "resize-down-indexer"
+  description = "The Cloud Function that resizes the node pool in GKE performing crawling (indexing-pool)"
 }
 
 # CI process should place the function in a different location
-variable index_resizer_function_code_path {
-    default = "/Users/david/development/repos/moneycol/infra/gke-resizer/code"
-    description = "Path to the code of the GKE resizer function for the indexing pool"
+variable "index_resizer_function_code_path" {
+  default     = "/Users/david/development/repos/moneycol/infra/gke-resizer/code"
+  description = "Path to the code of the GKE resizer function for the indexing pool"
 }
 
-variable crawler_resize_uri {
-    default = "https://europe-west1-moneycol.cloudfunctions.net/gke-resize"
-    description = "The URI of the resizer function to invoke"
+variable "crawler_resize_uri" {
+  default     = "https://europe-west1-moneycol.cloudfunctions.net/gke-resize"
+  description = "The URI of the resizer function to invoke"
 }
 
-variable crawler_resize_payload {
-    default = "{\"projectId\":  \"moneycol\",\"zone\": \"europe-west1-b\",\"clusterId\":  \"cluster-dev2\",\"nodePoolId\": \"indexing-pool\",\"nodeCount\":1}"
-    description = "The payload to resize"
+variable "crawler_resize_payload" {
+  default     = "{\"projectId\":  \"moneycol\",\"zone\": \"europe-west1-b\",\"clusterId\":  \"cluster-dev2\",\"nodePoolId\": \"indexing-pool\",\"nodeCount\":1}"
+  description = "The payload to resize"
 }
 
-variable crawler_resize_job_schedule {
-    # Indexing only happens on 1st of each month for now
-    default = "0 1 1 * *"
-    description = "The cron schedule to resize the indexing node pool to 1"
+variable "crawler_resize_job_schedule" {
+  # Indexing only happens on 1st of each month for now
+  default     = "0 1 1 * *"
+  description = "The cron schedule to resize the indexing node pool to 1"
 }
 
-variable crawling_done_pubsub_topic {
-    default = "dev.crawler.events"
-    description = "The resizer function (back to 0, crawling done) is triggered when message is published to this topic"
+variable "crawling_done_pubsub_topic" {
+  default     = "dev.crawler.events"
+  description = "The resizer function (back to 0, crawling done) is triggered when message is published to this topic"
 }
 
-variable indexer_sink_pubsub_topic {
-    default = "dev.moneycol.indexer.sink"
-    description = "The Pub/Sub topic/subscription where data ready to index is pushed by batcher/worker functions"
+variable "indexer_sink_pubsub_topic" {
+  default     = "dev.moneycol.indexer.sink"
+  description = "The Pub/Sub topic/subscription where data ready to index is pushed by batcher/worker functions"
 }

--- a/data-collector/crawler/infra/terraform/data-collector/versions.tf
+++ b/data-collector/crawler/infra/terraform/data-collector/versions.tf
@@ -1,14 +1,14 @@
 provider "google" {
-  project     = "moneycol"
-  region      = "europe-west1"
+  project = "moneycol"
+  region  = "europe-west1"
 }
 
 provider "google-beta" {
-  project     = "moneycol"
-  region      = "europe-west1"
+  project = "moneycol"
+  region  = "europe-west1"
 }
 
 provider "kubernetes" {
-  config_path    = "~/.kube/config"
+  config_path = "~/.kube/config"
   #config_context = "default"
 }

--- a/infra/gke/index.ts
+++ b/infra/gke/index.ts
@@ -15,9 +15,9 @@ const serverlessVpcConnectorMachineType = "f1-micro";
 // Create a GKE cluster
 const engineVersion = gcp.container.getEngineVersions({project: projectId, location: location}).then(v => v.latestMasterVersion);
 const cluster = new gcp.container.Cluster(name, {
-    name: name,
+    name,
     project: projectId,
-    location: location,
+    location,
     initialNodeCount: 1,
     minMasterVersion: engineVersion,
     removeDefaultNodePool: true,

